File: fmri_tools/segmentation/bias_field_correction.py
# -*- coding: utf-8 -*-

# python standard library inputs
import os

# external inputs
from nipype.interfaces.spm import NewSegment


def bias_field_correction(filename):
    """ Bias field correction

    Renzo recommended to perform a bias field correction using SPM12 before 
    doing the segmentation with FreeSurfer. FreeSurfer recommends Bias FWHM = 18 
    and Sampling distance = 2 for MEMPRAGE at 7 T, which is also set here. 
    Outputs are saved in the input folder.    

    Parameters
    ----------
    filename : str
        Path of input image.

    Returns
    -------
    None.

    Notes
    -------
    created by Daniel Haenelt
    Date created: 01-11-2018             
    Last modified: 12-10-2020
    
    """
       
    # get path of filename
    path = os.path.dirname(filename)

    # bias field correction
    os.chdir(path)
    bias = NewSegment()
    bias.inputs.channel_files = os.path.join(filename)
    bias.inputs.channel_info = (0.001, 18, (True, True))
    bias.inputs.affine_regularization = "mni"
    bias.inputs.sampling_distance = 2
    bias.inputs.use_v8struct = True
    bias.inputs.warping_regularization = [0, 0.001, 0.5, 0.05, 0.2]
    bias.inputs.write_deformation_fields = [False, False]
    bias.inputs.mfile = True
    tissue1 = ((os.path.join(pathSPM12, "tpm/TPM.nii"), 1), 1, (False,False), (False, False))
    tissue2 = ((os.path.join(pathSPM12, "tpm/TPM.nii"), 2), 1, (False,False), (False, False))
    tissue3 = ((os.path.join(pathSPM12, "tpm/TPM.nii"), 3), 2, (False,False), (False, False))
    tissue4 = ((os.path.join(pathSPM12, "tpm/TPM.nii"), 4), 3, (False,False), (False, False))
    tissue5 = ((os.path.join(pathSPM12, "tpm/TPM.nii"), 5), 4, (False,False), (False, False))
    tissue6 = ((os.path.join(pathSPM12, "tpm/TPM.nii"), 6), 2, (False,False), (False, False))
    bias.inputs.tissues = [tissue1, tissue2, tissue3, tissue4, tissue5, tissue6]
    bias.run()
    

File: fmri_tools/segmentation/bias_field_correction.m
function bias_field_correction(filename)
% Bias field correction.
%
% bias_field_correction(filename)
%
% Inputs:
%   filename - filename of input volume.
%
% Renzo recommended to perform a bias field correction using SPM12 before 
% doing the segmentation with FreeSurfer. FreeSurfer [1] recommends Bias 
% FWHM = 18 and Sampling distance = 2 for MEMPRAGE at 7 T, which is also 
% set here. Outputs are saved in the input folder. 
%
% [1] https://surfer.nmr.mgh.harvard.edu/fswiki/HighFieldRecon

% created by Daniel Haenelt
% Date created: 01-11-2018             
% Last modified: 07-01-2021

% get SPM12 root directory from matlab search path
pathSPM = what('spm12').path;

% set spm default parameters
spm('defaults','FMRI');
spm_get_defaults('stats.maxmem',2^35); % maxmen indicates how much memory can be used
spm_get_defaults('cmdline',true); % no gui

matlabbatch{1}.spm.spatial.preproc.channel.vols = {filename};
matlabbatch{1}.spm.spatial.preproc.channel.biasreg = 0.001;
matlabbatch{1}.spm.spatial.preproc.channel.biasfwhm = 18.0;
matlabbatch{1}.spm.spatial.preproc.channel.write = [1 1];
matlabbatch{1}.spm.spatial.preproc.tissue(1).tpm = {fullfile(pathSPM, 'tpm', 'TPM.nii,1')};
matlabbatch{1}.spm.spatial.preproc.tissue(1).ngaus = 1;
matlabbatch{1}.spm.spatial.preproc.tissue(1).native = [0 0];
matlabbatch{1}.spm.spatial.preproc.tissue(1).warped = [0 0];
matlabbatch{1}.spm.spatial.preproc.tissue(2).tpm = {fullfile(pathSPM, 'tpm', 'TPM.nii,2')};
matlabbatch{1}.spm.spatial.preproc.tissue(2).ngaus = 1;
matlabbatch{1}.spm.spatial.preproc.tissue(2).native = [0 0];
matlabbatch{1}.spm.spatial.preproc.tissue(2).warped = [0 0];
matlabbatch{1}.spm.spatial.preproc.tissue(3).tpm = {fullfile(pathSPM, 'tpm', 'TPM.nii,3')};
matlabbatch{1}.spm.spatial.preproc.tissue(3).ngaus = 2;
matlabbatch{1}.spm.spatial.preproc.tissue(3).native = [0 0];
matlabbatch{1}.spm.spatial.preproc.tissue(3).warped = [0 0];
matlabbatch{1}.spm.spatial.preproc.tissue(4).tpm = {fullfile(pathSPM, 'tpm', 'TPM.nii,4')};
matlabbatch{1}.spm.spatial.preproc.tissue(4).ngaus = 3;
matlabbatch{1}.spm.spatial.preproc.tissue(4).native = [0 0];
matlabbatch{1}.spm.spatial.preproc.tissue(4).warped = [0 0];
matlabbatch{1}.spm.spatial.preproc.tissue(5).tpm = {fullfile(pathSPM, 'tpm', 'TPM.nii,5')};
matlabbatch{1}.spm.spatial.preproc.tissue(5).ngaus = 4;
matlabbatch{1}.spm.spatial.preproc.tissue(5).native = [0 0];
matlabbatch{1}.spm.spatial.preproc.tissue(5).warped = [0 0];
matlabbatch{1}.spm.spatial.preproc.tissue(6).tpm = {fullfile(pathSPM, 'tpm', 'TPM.nii,6')};
matlabbatch{1}.spm.spatial.preproc.tissue(6).ngaus = 2;
matlabbatch{1}.spm.spatial.preproc.tissue(6).native = [0 0];
matlabbatch{1}.spm.spatial.preproc.tissue(6).warped = [0 0];
matlabbatch{1}.spm.spatial.preproc.warp.reg = [0.0 0.001 0.5 0.05 0.2];
matlabbatch{1}.spm.spatial.preproc.warp.affreg = 'mni';
matlabbatch{1}.spm.spatial.preproc.warp.samp = 2.0;
matlabbatch{1}.spm.spatial.preproc.warp.write = [0 0];

% run
spm_jobman('run', matlabbatch);
